util/gen_log_guide.ts

```diff
@@ -0,0 +1,503 @@
+        process.exit(-1);
+      }
+      if (!isLineType(lineType)) {
+        console.error(`Received invalid type specification: ${lineType ?? 'undefined'}`);
+        process.exit(-2);
+      }
+
+      const lineDoc = lineDocs[lineType];
+
+      if (!lineDoc) {
+        console.error(`Received missing type specification: ${lineType ?? 'undefined'}`);
+        process.exit(-3);
+      }
+
+      mappedLogLines[language] ??= [];
+      mappedLogLines[language]?.push(lineType);
+
+      const logRepo = new LogRepository();
+      let ret = '';
+      const lineDef = logDefinitions[lineType];
+      const structureNetworkArray = [
+        lineDef.type,
+        '2021-04-26T14:11:35.0000000-04:00',
+      ];
+      let lastIndex = 0;
+
+      for (const [name, index] of Object.entries(lineDef.fields)) {
+        if (['type', 'timestamp'].includes(name))
+          continue;
+        structureNetworkArray[index] = `[${name}]`;
+        lastIndex = Math.max(lastIndex, index);
+      }
+
+      for (let index = 2; index <= lastIndex; ++index)
+        structureNetworkArray[index] ??= '';
+
+      const structureNetwork = structureNetworkArray.join('|');
+      const structureLogLine = ParseLine.parse(logRepo, structureNetwork);
+      const structureLog = structureLogLine?.properCaseConvertedLine ??
+        structureLogLine?.convertedLine ?? '';
+
+      const examplesNetwork = lineDoc.examples[language]?.join('\n') ?? '';
+      const examplesLogLine = lineDoc.examples[language]?.map((e) => {
+        const line = ParseLine.parse(logRepo, e);
+        return line?.properCaseConvertedLine ?? line?.convertedLine;
+      }).join('\n') ?? '';
+
+      const regexes = lineDoc.regexes;
+
+      ret += `
+#### Structure
+
+\`\`\`log
+Network Log Line Structure:
+${structureNetwork}
+
+ACT Log Line Structure:
+${structureLog}
+\`\`\`
+`;
+
+      if (regexes) {
+        ret += `
+#### Regexes
+
+\`\`\`log
+Network Log Line Regex:
+${regexes.network}
+
+ACT Log Line Regex:
+${regexes.logLine}
+\`\`\`
+`;
+      }
+
+      ret += `
+#### Examples
+
+\`\`\`log
+Network Log Line Examples:
+${examplesNetwork}
+
+ACT Log Line Examples:
+${examplesLogLine}
+\`\`\`
+`;
+      return ret;
+    },
+  },
+};
+
+const enLogGuidePath = path.posix.relative(
+  curPath,
+  path.posix.join(curPath, 'docs', 'LogGuide.md'),
+);
+
+markdownMagic(
+  [
+    enLogGuidePath,
+    path.posix.relative(curPath, path.posix.join(curPath, 'docs', '*', 'LogGuide.md')),
+  ],
+  config,
+  (_error, output) => {
+    let exitCode = 0;
+    for (const lang of languages) {
+      // Only check this lang if it was converted/base file existed
+      if (
+        !(
+          output.some((o) =>
+            RegExp(('[^\\w]' + lang + '[^\\w]')).exec(o.originalPath.toLowerCase())
+          ) ||
+          (lang === 'en' && !output.some((o) => o.originalPath.includes(enLogGuidePath)))
+        )
+      )
+        continue;
```

Comment:
As a simplification suggestion, what about ignoring any language that has no mappedLogLines instead?

Reply:
I think maybe it makes sense to define the potential error states and what action to take (warning or error) for each state, then adjust the logic to match?

Here are the potential error states as I see them:

1. No language-specific markdown file found (e.g. no `docs/de/LogGuide.md` file)
  i. Currently ignored
2. Markdown file doesn't have every defined type in `lineDocs`
  i. Currently errors
3. Markdown file attempts to reference a type not defined in `lineDocs`
  i. Currently errors
4. Markdown file attempts to reference an invalid language
  i. Currently errors

If this is the behavior we want for these error states, then adjusting the logic as suggested breaks `2` since a `docs/de/LogGuide.md` file with no mapped lines referenced would not error.

Reply:
I guess I'm saying that it is unlikely that a markdown file would exist and have no translations.  Even though ignoring no mapped lines breaks a strict interpretation of (2), in practice I think the only case that we'll ever see is when translations exist and one is forgotten.  It's a much simpler condition to check correctly, which is why I suggest it, rather than future me trying to re-puzzle over this conditional.

Reply:
I have updated the logic used for checking for errors. The output looks like this if a blank LogGuide.md file is created in the zh-CN folder:

```
✔ docs\LogGuide.md Updated
 Transforms run
  ⁕ logLines:type=GameLog&lang=en
  ⁕ logLines:type=ChangeZone&lang=en
  ⁕ logLines:type=ChangedPlayer&lang=en
  ⁕ logLines:type=AddedCombatant&lang=en
  ⁕ logLines:type=RemovedCombatant&lang=en
  ⁕ logLines:type=partyList&lang=en
  ⁕ logLines:type=PlayerStats&lang=en
  ⁕ logLines:type=StartsUsing&lang=en
  ⁕ logLines:type=Ability&lang=en
  ⁕ logLines:type=networkCancelAbility&lang=en
  ⁕ logLines:type=networkDoT&lang=en
  ⁕ logLines:type=WasDefeated&lang=en
  ⁕ logLines:type=GainsEffect&lang=en
  ⁕ logLines:type=HeadMarker&lang=en
  ⁕ logLines:type=networkRaidMarker&lang=en
  ⁕ logLines:type=networkTargetMarker&lang=en
  ⁕ logLines:type=LosesEffect&lang=en
  ⁕ logLines:type=networkGauge&lang=en
  ⁕ logLines:type=ActorControl&lang=en
  ⁕ logLines:type=NameToggle&lang=en
  ⁕ logLines:type=Tether&lang=en
  ⁕ logLines:type=limitBreak&lang=en
  ⁕ logLines:type=StatusEffect&lang=en
  ⁕ logLines:type=networkUpdateHP&lang=en
  ⁕ TOC

Language cn is missing LogGuide doc entry for type GameLog
Language cn is missing LogGuide doc entry for type ChangeZone
Language cn is missing LogGuide doc entry for type ChangedPlayer
Language cn is missing LogGuide doc entry for type AddedCombatant
Language cn is missing LogGuide doc entry for type RemovedCombatant
Language cn is missing LogGuide doc entry for type partyList
Language cn is missing LogGuide doc entry for type PlayerStats
Language cn is missing LogGuide doc entry for type StartsUsing
Language cn is missing LogGuide doc entry for type Ability
Language cn is missing LogGuide doc entry for type networkCancelAbility
Language cn is missing LogGuide doc entry for type networkDoT
Language cn is missing LogGuide doc entry for type WasDefeated
Language cn is missing LogGuide doc entry for type GainsEffect
Language cn is missing LogGuide doc entry for type HeadMarker
Language cn is missing LogGuide doc entry for type networkRaidMarker
Language cn is missing LogGuide doc entry for type networkTargetMarker
Language cn is missing LogGuide doc entry for type LosesEffect
Language cn is missing LogGuide doc entry for type networkGauge
Language cn is missing LogGuide doc entry for type ActorControl
Language cn is missing LogGuide doc entry for type NameToggle
Language cn is missing LogGuide doc entry for type Tether
Language cn is missing LogGuide doc entry for type limitBreak
Language cn is missing LogGuide doc entry for type StatusEffect
Language cn is missing LogGuide doc entry for type networkUpdateHP
npm ERR! code ELIFECYCLE
npm ERR! errno 1
```

Hopefully this logic makes more sense?

Reply:
Sounds good, thanks!